storage/s3.go

```diff
@@ -382,6 +383,79 @@ func (s *S3) Get(
 	})
 }
 
+type SelectQuery struct {
+	ExpressionType  string
+	Expression      string
+	CompressionType string
+}
+
+func (s *S3) Select(ctx context.Context, url *url.URL, query *SelectQuery, resultCh chan<- json.RawMessage) error {
+	if s.dryRun {
+		return nil
+	}
+
+	input := &s3.SelectObjectContentInput{
+		Bucket:         aws.String(url.Bucket),
+		Key:            aws.String(url.Path),
+		ExpressionType: aws.String(query.ExpressionType),
+		Expression:     aws.String(query.Expression),
+		InputSerialization: &s3.InputSerialization{
+			CompressionType: aws.String(query.CompressionType),
+			JSON: &s3.JSONInput{
+				Type: aws.String("Lines"),
+			},
+		},
+		OutputSerialization: &s3.OutputSerialization{
+			JSON: &s3.JSONOutput{},
+		},
+	}
+
+	resp, err := s.api.SelectObjectContentWithContext(ctx, input)
+	if err != nil {
+		return err
+	}
+
+	reader, writer := io.Pipe()
+
+	go func() {
+		defer writer.Close()
+
+		eventch := resp.EventStream.Reader.Events()
+		defer resp.EventStream.Close()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, ok := <-eventch:
+				if !ok {
+					return
+				}
+
+				switch e := event.(type) {
+				case *s3.RecordsEvent:
+					writer.Write(e.Payload)
+				}
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(reader)
```

Comment:
do we really need to unmarshall the payload?

Reply:
The idea is that we need to find record delimiters, and the byte slices we get back from the `SelectObjectContent` operation don't align to record boundaries. This is a relatively straightforward way to find those boundaries and output the entire record at once, without having multiple concurrent select operations interleave their output.

Reply:
Good to know. Thanks.